### PR TITLE
Add Photon (Komoot) as fallback geocoding provider

### LINKED ISSUE #55: Add Photon (Komoot) as fallback geocoding provider when Nominatim is unavailable
## Background
The Nominatim API is blocked or unavailable in some geographic regions, causing all location searches to silently fail. We need a fallback geocoding provider.

## Research Summary
After testing all supported search patterns against multiple keyless APIs, **Photon (Komoot)** is the recommended fallback:
- Endpoint: `https://photon.komoot.io/api/?q={query}&limit=10&osm_tag=place`
- No API key required
- Independent infrastructure from Nominatim (AWS/Komoot-hosted)
- Passes all 11 search patterns: city, city+state, city+country, postal code, city+postal code, combos like `Birmingham, AL, US` and `London, Ontario, Canada`

Open-Meteo geocoding was evaluated and rejected — it treats the full query as a single token and returns zero results for comma-separated input.

## Implementation

**New models:**
- `Models/PhotonResult.cs` — GeoJSON FeatureCollection wrapper
- `Models/PhotonFeature.cs` — geometry.coordinates + properties

**Changes:**
- `GeocodingService.cs` — try Nominatim first, fall back to Photon on failure or empty results
- `WeatherJsonContext.cs` — register new models for AOT-safe JSON deserialization
- Add `ConvertPhotonResults()` mapper analogous to `ConvertNominatimResults()`

**Key notes:**
- Photon returns GeoJSON with coordinates as `[longitude, latitude]` — SWAP when mapping to GeocodingResult
- For postcode-only queries, omit `osm_tag=place` to allow postcode entries
- Map Photon properties to existing GeocodingResult: name, latitude, longitude, country, country_code, admin1 (state), admin2 (county)

## Acceptance Criteria
- Nominatim tried first; Photon used as fallback on error or empty results
- All existing search patterns work via Photon
- AOT/trimming compatibility maintained
- New tests cover the Photon conversion path

### PR DESCRIPTION
**Pull request by michaeljolley**
Closes #55

## Summary
Nominatim is blocked or unavailable in some regions, causing all location searches to silently fail there. This adds **Photon (Komoot)** as a fallback geocoding provider that activates when Nominatim returns an exception or empty results.

Photon was selected after live testing of all supported query patterns (research summary in #55). It is keyless, hosted on infrastructure independent of Nominatim, and handles every pattern correctly when called with `osm_tag=place` for free-text and without that filter for postcode-only queries.

## Changes
**New models** (`Microsoft.CmdPal.Ext.Weather.Models`):
- `PhotonResult` — GeoJSON FeatureCollection wrapper
- `PhotonFeature` / `PhotonGeometry` / `PhotonProperties` — feature shape

**Service changes** (`GeocodingService.cs`):
- `SearchWithProgressiveFallbackAsync` — at each iteration, if Nominatim throws or returns empty, tries Photon (`osm_tag=place`) for the same query before stripping the next comma segment.
- `SearchPostalCodeAsync` — every failure path (non-success status, null deserialization, empty list, exception) now falls through to Photon **without** `osm_tag=place` so postcode entries come through.
- New `SearchPhotonAsync(query, placesOnly, ct)` mirrors the Nominatim helper's logging + error patterns.
- New `ConvertPhotonResults` mapper translates GeoJSON to `GeocodingResult` with **explicit `[lon, lat]` → swap → `Latitude/Longitude`** (called out with an inline comment because it's a footgun) and the same skip rules as `ConvertNominatimResults`.

**JSON context** (`WeatherJsonContext.cs`):
- Registered `PhotonResult`, `PhotonFeature`, and `List<PhotonFeature>` for source-gen / AOT-safe deserialization. No reflection-based JSON anywhere.

**Tests** (`WeatherExtension.Tests/ConvertPhotonResultsTests.cs`, MSTest to match the rest of the suite):
14 tests covering:
- Basic field mapping (city-only, full properties)
- **Critical lon/lat swap** using asymmetric coordinates (10.5, 20.5) so a swap regression cannot pass
- Postcode entries (osm_value="postcode")
- Name fallback to City
- CountryCode uppercased
- OsmId null → Id=0
- Empty list → empty result
- Skip rules: null Geometry, single coordinate, null Coordinates, both Name+City null/whitespace
- Mixed valid+invalid input returns only the valid ones
- Multiple valid features preserved in order

## Notes
- Nominatim remains primary so its richer postcode handling is still preferred when available.
- The repo can only be fully built on Windows (cswinrt.exe is required for the WinRT codegen step). C# compilation phase was clean locally; CI on `windows-latest` is the authoritative build/test verdict.

🚬 *I love it when a plan comes together.*

— Hannibal, with B.A. on the wiring and Murdock on the test bench

### CODE REVIEW

**Comment by michaeljolley** [comment]
Both review comments are already addressed by commit `1f8c45e` (pushed a couple of minutes ago alongside the failing-test fix):

**Re: line 113 attempt cap** — The per-iteration `attempts++` is gone. `MaxFallbackAttempts` is now a shared `remainingCalls` budget that every Nominatim and Photon call decrements. The loop exits the moment the budget reaches zero, so total HTTP calls are guaranteed `<= MaxFallbackAttempts` regardless of how many providers an iteration touches. This is what makes the `...CapsAttemptsAtMaxFallbackAttempts` test green again (3 total calls for the 5-segment input instead of 6).

**Re: line 122 cancellation before Photon** — Added `ct.ThrowIfCancellationRequested();` immediately before `SearchPhotonAsync` is invoked. A cancellation requested mid-Nominatim now bails out before any Photon request leaves the box.

Same commit also marks Nominatim as dead after it throws, so we don't burn shared-budget slots on a provider that's clearly unreachable.

**Review by michaeljolley — COMMENTED** [review]
**Squad QA Review — Amy Allen (Lead) + Murdock (Test Engineer)**

✅ **Both veto-capable reviewers approve this PR.**

**Amy Allen (Team Lead):**
- GeoJSON [lon, lat] → [lat, lon] swap correctly implemented
- Fallback budget (MaxFallbackAttempts) prevents runaway HTTP calls
- AOT-safe JSON registration in WeatherJsonContext
- osm_tag=place filter logic handles all 11 query patterns correctly

**Murdock (QA/Test Engineer):**
- 15 edge cases in ConvertPhotonResultsTests: null geometry, empty features, missing properties, coordinate swap with asymmetric values (10.5, 20.5)
- Fallback budget test caught a real CI regression
- AOT wiring verified
- Non-blocking: full Nominatim→Photon integration test tracked in #59

**Verdict: Ready for Michael's review and merge.** 🎩

File: WeatherExtension/Models/PhotonFeature.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Bald Bearded Builder LLC
+// Bald Bearded Builder LLC licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.CmdPal.Ext.Weather.Models;
+
+public sealed class PhotonFeature
```
Comment (michaeljolley) [inline WeatherExtension/Models/PhotonFeature.cs]: These should probably be internal rather than public right?

File: WeatherExtension/Models/PhotonResult.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Bald Bearded Builder LLC
+// Bald Bearded Builder LLC licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.CmdPal.Ext.Weather.Models;
+
+public sealed class PhotonResult
```
Comment (michaeljolley) [inline WeatherExtension/Models/PhotonResult.cs]: Same. Internal vs public?
